src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignerInfo.cs

```diff
@@ -591,100 +578,236 @@ private bool CheckHash(bool compatMode)
-                            }
+            // Some CMS implementations exist which do not sort the attributes prior to
+            // generating the signature.  While they are not, technically, validly signed,
+            // Windows and OpenSSL both support trying in the document order rather than
+            // a sorted order.  To accomplish this we will build as a SEQUENCE OF, but feed
+            // the SET OF into the hasher.
+            using (compatMode ? writer.PushSequence() : writer.PushSetOf())
+            {
+                foreach (AttributeAsn attr in signedAttributes)
+                {
+                    attr.Encode(writer);
 
-                            var digestAttr = (Pkcs9MessageDigest)obj.Values[0];
+                    // .NET Framework doesn't seem to validate the content type attribute,
+                    // so we won't, either.
 
-                            if (!contentDigest.AsSpan().SequenceEqual(digestAttr.MessageDigest))
-                            {
-                                throw new CryptographicException(SR.Cryptography_BadHashValue);
-                            }
+                    if (attr.AttrType == Oids.MessageDigest)
+                    {
+                        CryptographicAttributeObject obj = PkcsHelpers.MakeAttribute(attr);
 
-                            invalid = false;
+                        if (obj.Values.Count != 1)
+                        {
+                            throw new CryptographicException(SR.Cryptography_BadHashValue);
                         }
-                    }
 
-                    if (compatMode)
-                    {
-                        writer.PopSequence();
+                        var digestAttr = (Pkcs9MessageDigest)obj.Values[0];
 
-                        byte[] encoded = writer.Encode();
-                        encoded[0] = 0x31;
-                        hasher.AppendData(encoded);
+                        if (!digest.SequenceEqual(digestAttr.MessageDigest))
+                        {
+                            throw new CryptographicException(SR.Cryptography_BadHashValue);
+                        }
+
+                        hasMatchingDigestAttr = true;
                     }
-                    else
+                    else if (attr.AttrType == Oids.ContentType)
                     {
-                        writer.PopSetOf();
+                        hasContentAttr = true;
+                    }
+                }
+            }
+
+            // Message-digest is required when signed attributes are present.
+            if (!hasMatchingDigestAttr || (needsContentAttr && !hasContentAttr))
+            {
+                throw new CryptographicException(SR.Cryptography_Cms_MissingAuthenticatedAttribute);
+            }
 
+            if (compatMode)
+            {
+                byte[] encoded = writer.Encode();
+                encoded[0] = 0x31;
+                return verify(encoded);
```

Comment:
VerifyCallback takes a span, I believe; so we could (if we wanted to) use a stackbuf or pooled array here.

I'm fine with saying it's not worth the effort.